File: java/interpreter/src/com/miwan/interpreter/lexical/LexStream.java
package com.miwan.interpreter.lexical;

import java.util.List;
import java.util.function.Function;

public class LexStream {
	public LexStream(List<Lexeme> content) {
		this.content = content;
	}

	public Lexeme eat() {
		if (pos >= this.content.size())
			return null;
		return content.get(pos++);
	}

	public Lexeme current() {
		if (pos >= this.content.size())
			return null;
		return content.get(pos);
	}

	public Lexeme peek(int c) {
		if (pos + c >= content.size())
			return null;
		return content.get(pos + c);
	}

	public Lexeme peek() {
		return peek(1);
	}

	public boolean hasNext() {
		return peek() != null;
	}

	public static boolean test(Lexeme lex, Function<Lexeme, Boolean> tester) {
		if (lex != null)
			return tester.apply(lex);
		return false;
	}

	@Override
	public String toString() {
		return current().text;
	}

	private int pos = 0;
	final private List<Lexeme> content;
}

File: cxx/interpreter/src/lex_stream.cpp
#include <exception>
#include <lexical/lex_stream.h>
#include <utility>

interpreter::lex_stream::lex_stream(auto &&_content)
    : content(std::forward<decltype(_content)>(_content)), current_index(0) {}
interpreter::lex_stream::lex_stream(interpreter::lex_stream &&) = default;
interpreter::lex_stream::lex_stream(const interpreter::lex_stream &) = default;

interpreter::lexeme interpreter::lex_stream::eat() {
  if (this->current_index >= this->content.size()) {
    throw std::runtime_error("out of range");
  }
  return content[current_index++];
}

interpreter::lexeme interpreter::lex_stream::current() {
  if (this->current_index >= this->content.size()) {
    throw std::runtime_error("out of range");
  }
  return content[current_index];
}

interpreter::lexeme interpreter::lex_stream::peek() { this->peek(1); }

interpreter::lexeme interpreter::lex_stream::peek(std::size_t c) {
  if (this->current_index + c >= content.size()) {
    throw std::runtime_error("out of range");
  }
  return content[current_index + c];
}

bool interpreter::lex_stream::has_next() {
  return this->current_index < content.size() - 1;
}
